Add ability to get driver from NVIDIA's official RHEL repository, to avoid version issues

Aietes opened the pull request:
On certain distros, where the NVIDIA driver is typically installed via the NVIDIA driver repositories, trying to install the driver from the download site will commonly fail with a 404 error, like described in #170 and #124 and #67.

There have been approaches to solve this, e.g. in #184, but this will not work reliably. The key problem is that NVIDIA is not providing consistent versions of their drivers on different distribution channels. For Rocky Linux, and all other RHEL distros, the latest version on the download site differs. e.g. at the time of writing, the latest on downloads `.run` is 570.124.**04**, while the `.rpm` version in the RHEL repository is 570.124.**06**:

Driver download: https://download.nvidia.com/XFree86/Linux-x86_64/570.124.04/
RHEL repository: https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/nvidia-driver-libs-570.124.06-1.el9.x86_64.rpm

Using a different version, even if it's a small difference in the minor version, does not work in my experience. One could always manually install a driver, but in the case of RHEL it's way better to get the NVIDIA driver installed and updated via the default package manager and update cycle.

To solve this issue, I've created a [fork](https://github.com/Aietes/nixGL/tree/rhel) with an additional derivation in nixGL, that gets the driver from the RHEL repository from the `.rpm`. This works well for me, and should solve the issue for RHEL users. It it would be nice if it could be merged.

To get the driver from the RHEL `.rpm` instead of the driver download site, simply set `driverSource` to `rhel`. Also ensure that `enable32bits` is changed from its default to `false`, as at this point I did not include the 32bit version of the driver. Here's an example from my `.nix` using `home-manager`:

```nix
{
  nixpkgs = {
    overlays = specialArgs.overlays;
    config = { allowUnfree = true; };
  };

  home.packages = with pkgs; [
    (nixgl.override {
      driverSource = "rhel";
      enable32bits = false;
    }).auto.nixGLDefault
    (config.lib.nixGL.wrap ghostty)
  ];

  nixGL = {
   packages = pkgs.nixgl.override {
    driverSource = "rhel";
    enable32bits = false;
   };
   defaultWrapper = "nvidia";
  };
```

If you face the driver problem and you are on any RHEL distro like Rocky, Fedora, etc., you can use the nixGL overlay from my fork until this is merged:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    nixgl = {
      url = "github:Aietes/nixGL/rhel";
    };
  };
  ....
```

Something similar can probably be done for Debian and other distros, using the official NVIDIA driver for a specific distro as the source for nixGL.

Fixes #170 #124 #67

Aietes committed:
feat: option to get driver from rpm from NVIDIA RHEL repository

Aietes committed:
refactor: remove driver RPM, as it's not really needed for nixGL to work

Aietes committed:
fix(rhel): add libnvidia-ml as it is required by btop and other apps

copy additional rpm to temp in unpack phase

Aietes committed:
fix(rhel): make rhel the condition instead of driver

Aietes committed:
docs(readme): update README with instruction to get driver from RHEL repository

Aietes commented:
Since it seems that the driver version is consistent across Debian, Fedora and RHEL, at lease at this point, pulling the driver from the RHEL repository very likely also works on other distros, pending somebody testing this.

dyfrgi commented on nixGL.nix:
```diff
@@ -51,25 +55,81 @@ let
         }
-        export LD_LIBRARY_PATH=${lib.makeLibraryPath mesa-drivers}:${lib.makeSearchPathOutput "lib" "lib/vdpau" libvdpau}:${glxindirect}/lib:${lib.makeLibraryPath [libglvnd]}"''${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
+        export LIBVA_DRIVERS_PATH=${
+          lib.makeSearchPathOutput "out" "lib/dri" (mesa-drivers ++ vadrivers)
```
Could you move the formatting changes here, and elsewhere in the PR, into their own commit?

Aietes replied:
Sorry for those, autoformatting...

dyfrgi reviewed (changes requested):
This looks fine to me as an approach to implementing the idea, but it's strange to me that it's necessary. Are the RHEL repositories more reliable? Have you inquired on the NVidia developer forums about this strange behavior of deleting old versions from the download site? Does the license permit redistribution? If so, we could also consider making our own repository of them.

Aietes commented:
I'm short on time right now, but will look into tests when I find time. Do you have a specific testing methodology?

dyfrgi commented:
I don't have a testing method, no. The testing in this project needs some work in general. There's all.nix and Test.hs today, neither of which quite fits. I think it's fine to merge this with only manual testing, especially since I don't really have any suggestions on where to hang new tests for this.

> This is needed because the specific driver version used in RHEL rarely (never) matches a version available in the NVIDIA repos

Ohhh, I think I misunderstood what you meant by "consistent version across distribution channels" in the README.md change. I thought you meant that it was the same number but the files were meaningfully different. If they're always different versions, I wonder if they could be merged into one auto detection lists in the future?

With the formatting changes stripped out, I think this would be good to merge. I'd like to see either an automated test or another report from a user about this working for them, including the update script.

Aietes committed:
fix(rhel): support rhel major version option

Aietes committed:
fix(rhel): select NVML package by driver branch

NVIDIA dropped the standalone libnvidia-ml RPM with the 610 driver branch
and moved libnvidia-ml.so into the new nvidia-driver-common package. Pick
the providing package based on the detected driver branch so RHEL driver
extraction keeps working on 610+ (e.g. Rocky Linux 10) as well as 580..595.

Aietes committed:
Merge remote-tracking branch 'upstream/main' into rhel

# Conflicts:
#	nixGL.nix

Aietes committed:
style(rhel): drop unrelated reformatting, minimize diff vs main

The original RHEL feature commits ran nixfmt over the whole file, mixing
~70 lines of unrelated reformatting into the diff (the maintainer asked
for these to be stripped before merge). Rebuild nixGL.nix and README.md
from upstream and re-apply only the RHEL feature, so the diff contains
only the new functionality.

Aietes committed:
docs(rhel): clarify why the rhel driver source is needed

The two NVIDIA distribution channels (XFree86 .run installers vs. the RHEL
CUDA repo) carry different sets of version numbers, so the installed driver
version on a RHEL box often has no matching .run. Spell this out so the
rationale is clear from the code.

Aietes committed:
fix(rhel): address review feedback on the rhel driver source

- Fail fast with a clear message when driverSource = "rhel" is combined
  with enable32bits = true (the RPM path ships 64-bit libraries only),
  instead of a confusing missing-attribute error later.
- Replace the null nvidiaDrivers placeholder with a lazy throw explaining
  that the Bumblebee wrapper is unavailable with the RPM-only libs path.
- Set dontInstall = true on nvidiaFromRpm to make the no-install intent
  explicit (buildPhase populates $out; there is no Makefile).
- Drop Fedora from the RHEL docs/heading: NVIDIA serves Fedora from a
  separate repo path, so the rhelN URLs don't target it.

Aietes commented on nixGL.nix:
```diff
@@ -69,7 +83,64 @@ let
+
+      # Extract the NVIDIA shared libraries from the RPMs.
+      nvidiaFromRpm = stdenv.mkDerivation {
+        pname = "nvidia";
+        name = "nvidia-${version}-nixGL-rpm";
+        inherit version;
+        preferLocalBuild = true;
+        allowSubstitutes = false;
+        nativeBuildInputs = [ rpm cpio ];
+        srcs = [ rpmLibs rpmMl ];
```
Addressed in b63a594 — added `dontInstall = true;`. In practice the default `installPhase` was already a no-op here (no Makefile in the build dir), but this makes the intent explicit.

Aietes commented on nixGL.nix:
```diff
@@ -69,7 +83,64 @@ let
+          # the driver branch) only ships libraries under usr/lib64.
+          rpm2cpio nvidia-ml.rpm | cpio -idmv
+          mv usr/lib64/* $out/lib
+        '';
+      };
+
+      nvidiaDrivers = if driverSource == "rhel" then
+        null
+      else
+        (linuxPackages.nvidia_x11.override { }).overrideAttrs
```
Addressed in b63a594 — replaced the `null` with a lazy `throw` that explains the Bumblebee wrapper is unavailable on the RPM-only libs path. It is only forced if `nixGLNvidiaBumblebee` is built; `nixGLNvidia`/`nixVulkanNvidia` and the `auto.*` wrappers still evaluate fine.

Aietes commented on README.md:
````diff
@@ -134,6 +134,36 @@ Then, use the flake's `overlay` attr:
 }
 ```
 
+### Install from RHEL repository (Rocky, CentOS, Fedora, etc.)
````
Addressed in b63a594 — dropped Fedora from the heading and the code comment, since NVIDIA serves Fedora from a separate repo path that the `rhelN` URLs do not target.

Aietes commented on nixGL.nix:
```diff
@@ -9,14 +9,28 @@ nvidiaHash ? null,
 # Enable 32 bits driver
 # This is one by default, you can switch it to off if you want to reduce a
 # bit the size of nixGL closure.
 enable32bits ? stdenv.hostPlatform.isx86
 , stdenv, writeTextFile, shellcheck, pcre, runCommand, linuxPackages
 , fetchurl, lib, runtimeShell, bumblebee, libglvnd, vulkan-validation-layers
 , mesa, libvdpau-va-gl, intel-media-driver, pkgsi686Linux, driversi686Linux
-, zlib, libdrm, xorg, wayland, gcc, zstd }:
+, zlib, libdrm, xorg, wayland, gcc, zstd, rpm, cpio }:
 
 let
```
Addressed in b63a594 — added a fail-fast `assert` with a clear message when `driverSource = "rhel"` is combined with `enable32bits = true`.

dyfrgi reviewed (approved):
Are the ml libraries still in the .run file for non-rhel sources? If not we should probably document that, though it doesn't have to be done in this PR.
